tests/conftest.py

```diff
@@ -1,25 +1,152 @@
+        ]
+        self.db_ports = {"5432/tcp": self.config.PSQL_PORT}
+
+    def get_postgresql_image(self):
+        """Output the PostgreSQL image from the configuration.
+        Returns:
+            str: The PostgreSQL image name and version tag.
+        """
+        return "{}:{}".format(self.config.IMAGE_NAME, self.config.IMAGE_VERSION)
+
+    def start_container(self):
+        """Start PostgreSQL Container."""
+        if self.get_db_if_running():
+            return
+
+        try:
+            self.docker_client.images.pull(self.get_postgresql_image())
+        except Exception:
+            logging.error("Failed to pull postgres image")
+            raise RuntimeError
+
+        self.container = self.docker_client.containers.run(
+            self.get_postgresql_image(),
+            detach=True,
+            auto_remove=True,
+            name=self.config.CONTAINER_NAME,
+            environment=self.db_environment,
+            ports=self.db_ports,
+        )
+        logging.info("PostgreSQL container running!")
+
+        apply_migrations()
 
-    return app 
+    def stop_if_running(self):
+        try:
+            running = self.docker_client.containers.get(self.config.CONTAINER_NAME)
+            logging.info(f"Killing running container '{self.config.CONTAINER_NAME}'")
+            running.stop()
+        except Exception as e:
+            if "404 Client Error: Not Found" in str(e):
+                return
+            raise e
+
+    def get_db_if_running(self):
+        """
+        Return the container or None.
+        """
+        try:
+            return self.docker_client.containers.get(self.config.CONTAINER_NAME)
+        except Exception as e:
+            if "404 Client Error: Not Found" in str(e):
+                return
+
+
+def apply_migrations():
+    """Apply Database Migrations
+    Applies the database migrations to the test database container.
+    Args:
+        config (object): Configuration object to pull the needed components from.
+    """
+    applied_migrations = False
+    retries = 0
+    max_retries = 3
+
+    with testing_app.app_context():
+        relative_path = os.path.dirname(os.path.relpath(__file__))
+        absolute_path = os.path.dirname(os.path.abspath(__file__))
+        root_path = absolute_path.split(relative_path)[0]
+        migrations_dir = os.path.join(
+            root_path, 'migrations')
+
+        while retries < max_retries and applied_migrations is False:
+            logging.info('Attempting to apply migrations ({} of {})...'.format(
+                retries + 1, max_retries))
+            try:
+                upgrade(directory=migrations_dir)
+                applied_migrations = True
+            except Exception:
+                retries += 1
+                sleep(2)
+
+
+@pytest.fixture(scope="session", autouse=True)
+def database():
+    postgres = PostgreSQLContainer()
+    postgres.start_container()
+
+    yield db
+
+    postgres.stop_if_running()
```

Comment:
I believe this should get called if you exit tests with ctrl+c fyi, unless you terminate it I believe.

Reply:
yeah, it gets called with `ctrl + c`, which actually makes life pretty easy, since you do not end up with lingering containers.